gh-120284: Enhance asyncio.Runner's run() method to accept more object types

ronf opened the pull request:
This commit makes asyncio.Runner's run() method accept the same object types as asyncio.run_until_complete(). Instead of accepting only a coroutine, run() will now accept an awaitable, coroutine, asyncio.Future, or asyncio.Task.

<!-- gh-issue-number: gh-120284 -->
* Issue: gh-120284
<!-- /gh-issue-number -->

ronf committed:
Enhance asyncio.Runner's run() method to accept more object types

This commit makes asyncio.Runner's run() method accept the same object
types as asyncio.run_until_complete(). Instead of accepting only a
coroutine, run() will now accept an awaitable, coroutine,
asyncio.Future, or asyncio.Task.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

kumaraditya303 commented on Lib/asyncio/runners.py:
```diff
@@ -97,7 +103,21 @@ def run(self, coro, *, context=None):
             context = self._context
-        task = self._loop.create_task(coro, context=context)
+
+        if futures.isfuture(coro):
```
You replace most of this with call to ensure_future.

ronf replied:
That's what I did originally, but it caused problems with some of the unit tests. In particular, ensure_future() had a call to get_event_loop() in it which caused a unit test to fail. Also, in my original change, I needed to modify ensure_future() to take a context argument and that changed a public API, which caused other associated unit test issues. More importantly, ensure_future() also had code which called close() on the coro or future, which isn't wanted here.

After trying out various alternatives, I settled on just taking the few lines from ensure_future() that were needed and modifying them to include only what this new code required. It was a smaller change, and in my opinion lower risk and more readable.

kumaraditya303 replied:
Maybe extract it to a new function? I really don't want to repeat the logic of awaitable trick here.

ronf replied:
I'll take another look at this tonight.

ronf replied:
Ok - I figured out a way to leverage the code in ensure_future() for this - see commit 4ed7357.

One thing I'm not thrilled about here is that this adds a new optional keyword argument to `ensure_future()`, which is a public API. I could move the entire logic of `ensure_future()` into a new function instead to avoid changing its signature, but that's kind of ugly in a different way.

One other thing to note is that this now raises a `TypeError` instead of a `ValueError` when a caller passes in a `coro` which is not one of the expected types, since that's what `ensure_future()` raised.

kumaraditya303 commented on Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst:
```diff
@@ -0,0 +1 @@
+Make asyncio.Runner's run() method accept the same object types as asyncio.run_until_complete(). Instead of accepting only a coroutine, run() will now accept an awaitable, coroutine, asyncio.Future, or asyncio.Task.
```
use rst markup please

ronf replied:
Done, though I wasn't sure how to validate that the references will work.

ronf committed:
Add RST markup to NEWS entry.

ronf committed:
Fix an incorrect reference in NEWS entry.

Eclips4 commented on Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst:
```diff
@@ -0,0 +1,4 @@
+Enhance :meth:`asyncio.Runner.run` to accept the same object types as
+:meth:`asyncio.BaseEventLoop.run_until_complete`. Instead of accepting
+only a coroutine, it now accepts an awaitable, coroutine,
```
```suggestion
only a :term:`coroutine`, it now accepts an :term:`awaitable`, :term:`coroutine`,
```

ronf replied:
I'd be happy to make this change, but I'm still running into an issue with the reference to asyncio.BaseEventLoop.run_until_complete not being accepted:

`/home/runner/work/cpython/cpython/build/NEWS:183: WARNING: py:meth reference target not found: asyncio.BaseEventLoop.run_until_complete`

Any suggestions for fixing that?

kumaraditya303 replied:
Fixed it

ronf replied:
Thanks - I have also added the change requested above as commit df87f52.

itamaro commented on Lib/asyncio/runners.py:
```diff
@@ -97,7 +103,21 @@ def run(self, coro, *, context=None):
+            task = coro
+        elif coroutines.iscoroutine(coro):
+            task = self._loop.create_task(coro, context=context)
+        elif inspect.isawaitable(coro):
```
is there a more lightweight to check if coro is an awaitable?
just importing `inspect` is pretty heavy - perhaps defer the import only to when it's needed? (i.e. if coro is neither a coroutine nor a future nor a task)

ronf replied:
There's already an import of "inspect" in Lib/asyncio/tasks.py even before my changes here, and that's imported by Lib/asyncio/runners.py, so I don't think trying to defer the import makes sense here. If I'm able to factor out the code I need from tasks.py into a common function, I may be able to avoid the direct import in runners.py and have only the import that already existed in tasks.py, but either way the inspect module is going to be imported.

The function `inspect.isawaitable()` also doesn't look all that expensive to run:

```python
def isawaitable(object):
    """Return true if object can be passed to an ``await`` expression."""
    return (isinstance(object, types.CoroutineType) or
            isinstance(object, types.GeneratorType) and
                bool(object.gi_code.co_flags & CO_ITERABLE_COROUTINE) or
            isinstance(object, collections.abc.Awaitable))
```

itamaro replied:
you're right, this wouldn't be a regression since `inspect` is already imported in `tasks.py`.
the specific function may be inexpensive -- it's the overall import of the `inspect` module that's quite heavy.
we could try avoiding it in `tasks.py` too, but that would be a separate issue and PR.

ronf replied:
Thanks - makes sense.

kumaraditya303 commented on Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst:
```diff
@@ -0,0 +1,4 @@
+Enhance :meth:`asyncio.Runner.run` to accept the same object types as
+:meth:`asyncio.BaseEventLoop.run_until_complete`. Instead of accepting
```
```suggestion
:meth:`~asyncio.loop.run_until_complete`. Instead of accepting
```

kumaraditya303 committed:
Update Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst

ronf committed:
Add term markup for awaitable and coroutine in NEWS item

ronf committed:
Avoid duplicating code from ensure_future in Runner.run

willingc commented on Lib/asyncio/runners.py:
```diff
@@ -104,20 +102,7 @@ def run(self, coro, *, context=None):
-            raise ValueError("An asyncio Future, asyncio.Task, "
-                             "awaitable, or coroutine was expected, "
-                             "got {!r}".format(coro))
+        task = tasks.ensure_future(coro, loop=self._loop, context=context)
```
I'm not sure that we should use `ensure_future` since the docs flag it as deprecated since 3.10.

ronf replied:
I don't think the function is deprecated. According to the docs, 3.10 only deprecates not specifying a loop explicitly when there's no currently running event loop.

In the specific case we're using it here, loop will always be specified explicitly, as it will be the loop associated with the `asyncio.Runner` instance that `run()` is being called on.

willingc reviewed (commented):
Thanks for the PR.

Jason-Y-Z commented on Lib/asyncio/tasks.py:
```diff
@@ -745,7 +745,10 @@ async def _wrap_awaitable(awaitable):
+        if context is None:
+            return loop.create_task(coro_or_future)
+        else:
+            return loop.create_task(coro_or_future, context=context)
```
If I read the [definition of `create_task`](https://github.com/python/cpython/blob/95f5c89b545beaafad73f05a695742da3e90bc41/Lib/asyncio/tasks.py#L402) correctly, I think we can pass in `context` directly, since we also do a `None` check in `create_task` (unless I missed something).

ronf replied:
Removing the if/else here causes a couple of test failures:

2 tests failed:
    test.test_asyncio.test_base_events test_ssl

The version of create_task being called here is actually the one in base_events.py (line 462), and the problem is that the current unit tests define their own EventLoop class which supports only the old API, without a context argument. For example, in test_base_events, there's the following:

```python
    def test_create_task(self):
        class MyTask(asyncio.Task):
            pass

        async def test():
            pass

        class EventLoop(base_events.BaseEventLoop):
            def create_task(self, coro):
                return MyTask(coro, loop=loop)

        loop = EventLoop()
        self.set_event_loop(loop)

        coro = test()
        task = asyncio.ensure_future(coro, loop=loop)
        self.assertIsInstance(task, MyTask)

        # make warnings quiet
        task._log_destroy_pending = False
        coro.close()
```

This test could be updated to support the new API, but this exact issue could also appear in application code which defines its own event loop based on the old API. So, I think the safest thing to do is to only call the new API when a context is actually passed in.

Jason-Y-Z replied:
Make sense, thanks for clarifying!

Jason-Y-Z reviewed (commented):
Thanks for the PR! Overall logic makes sense to me, left a small comment

picnixz commented on Lib/asyncio/runners.py:
```diff
@@ -84,9 +84,13 @@ def get_loop(self):
+        If the argument is a coroutine, it is wrapped in a Task.
+
+        Return the coro's result or raise an exception.
+        """
```
```suggestion
        The argument can be any awaitable object.
        
        If the argument is a coroutine, it is wrapped in a Task.

        Return the awaitable's result or raise an exception.
        """
```

I know that the parameter is called "coro" but I'd prefer not using "coro's" in the docstring.

picnixz commented on Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst:
```diff
@@ -0,0 +1,4 @@
+Enhance :meth:`asyncio.Runner.run` to accept the same object types as
+:meth:`~asyncio.loop.run_until_complete`. Instead of accepting
+only a :term:`coroutine`, it now accepts an :term:`awaitable`,
+:term:`coroutine`, :class:`asyncio.Future`, or :class:`asyncio.Task`.
```
```suggestion
Allow :meth:`asyncio.Runner.run` to accept :term:`awaitable`
objects instead of simply :term:`coroutine`\s.
```

You can simplify the description like this. You should also change the RST of `asyncio.Runner.run` (not just the docstring). You can also create a true What's New entry since it's good to highlight this change IMO.

picnixz commented on Lib/asyncio/tasks.py:
```diff
@@ -720,7 +720,7 @@ async def sleep(delay, result=None):
 
 
-def ensure_future(coro_or_future, *, loop=None):
+def ensure_future(coro_or_future, *, loop=None, context=None):
```
You should document this new parameter in the online docs.

ronf replied:
Done, in commit e9e7c29.

picnixz reviewed (commented):
Some nitpicks (feel free to argue).

ronf committed:
Accept proposed doc change for asyncio/runners.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ronf committed:
Accept proposed change in  Misc/NEWS.d/next/Library/2024-06-15-23-38-36.gh-issue-120284.HwsAtY.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ronf committed:
Document new context argument in ensure_future()

ronf committed:
Remove trailing whitespace.

picnixz commented on Doc/library/asyncio-future.rst:
```diff
@@ -49,6 +49,9 @@ Future Functions
 
+   An optional keyword-only *context* argument allows specifying a
+   custom :class:`contextvars.Context` to use when creating a new task.
+
```
Maybe mention that the current context is used if none is specified (I think the other functions/methods that support contextvars do it).

ronf replied:
From what I can tell, `ensure_future()` sets no context at all if you pass in `None` to it. It calls `create_task()` on the current event loop with no context argument in that case. What happens from there would depend on the event loop being used.

I do see code in the `Task()` constructor which calls `contextvars.copy_context()` if you pass `None` in as a context and that's what event loops usually call.  I'm hesitant to claim that behavior at the `ensure_future()` level, though, given the multiple layers in between and the possibility for the event loops to do something different.

Looking at `Runner.run()`, it passes in the runner's default context if no context is specified when `Runner.run()` is called, and it is documented as such. That default context seems to be set to a copy of the context active at the time the Runner object is first used (during  its lazy init). That then gets passed explicitly to `ensure_future()` if no context is passed into `Runner.run()`.

Note that the runner's default context may not always be the same as the context which is active at the time `Runner.run()` is called.

picnixz replied:
You're entirely right. I incorrectly assumed that it would be like `Runner.run()` but that's not what the code actually does. Maybe we could expand the docs by saying:

```diff
-custom :class:`contextvars.Context` to use when creating a new task.
+custom :class:`contextvars.Context` to use when creating a new task
+via :meth:`loop.create_task <asyncio.loop.create_task>`.
```

ronf replied:
I don't think the loop version of the create_task() method is a public function. There's only asyncio.create_task() (and more recently TaskGroup.create_task), but those are not used by Runner.

picnixz replied:
Then I really wonder how we can teach the None argument in that case. I don't have any good suggestion but we can also decide to be deliberately imprecise... I'll leave it to another reviewer :/

ronf replied:
I looked this over again, and it turns out that the same indirection happening with `ensure_future()` calling into the internal `create_task()` on the event loop also happens in `asyncio.create_task()`, and yet it documents context as:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` for the *coro* to run in.
   The current context copy is created when no *context* is provided.
```

Given that both of these are likely to call the constructor for `Task` which does the actual copy and documents that:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` for the *coro* to run in.
   If no *context* is provided, the Task copies the current context
   and later runs its coroutine in the copied context.
```

So, perhaps it's ok after all to document context in `ensure_future()` as something like:

```
   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` to use when creating a new task.
   A copy of the current context is used when no *context* is provided.
```

There's a chance that a custom event loop could fail to pass the context through properly and change this behavior but that's fairly unlikely (and might in fact be considered a bug).

ronf committed:
Specify what happens in ensure_future with a context of None.

1st1 commented on Doc/library/asyncio-future.rst:
```diff
@@ -64,6 +68,8 @@ Future Functions
       Deprecation warning is emitted if *obj* is not a Future-like object
       and *loop* is not specified and there is no running event loop.
 
+   .. versionchanged:: 3.12
```
```suggestion
   .. versionchanged:: 3.13
```

1st1 commented on Lib/asyncio/tasks.py:
```diff
@@ -720,7 +720,7 @@ async def sleep(delay, result=None):
 
 
-def ensure_future(coro_or_future, *, loop=None):
+def ensure_future(coro_or_future, *, loop=None, context=None):
```
Sadly this change is a no-go.

If `coro_or_future` is a future or task, the `context` argument would be ignored, breaking the future user assumption that the context they pass would actually be applied.

I suggest a different approach to this PR:

`Runner.run(o)` should check if `o` is a coroutine -- if it is, keep things as is. If not, it should wrap it in an ad-hoc "proxy" coroutine `async def proxy(o=o): return await o` and pass `proxy` to `create_task()`.

ronf replied:
I have made the requested changes; please review again.

ronf replied:
(apologies Yury for misspelling your first name in the last commit!)

1st1 reviewed (dismissed):
The proposed implementation makes `ensure_future` unpredictable so I suggest taking a different route.

ronf committed:
Rework solution to address Yuri's review comments

This version has reverted the change in ensure_future, and only
changes Runner.run().

kumaraditya303 reviewed (changes requested):
Looks fine but the docs also need to be updated for this.

ronf committed:
Update documentation in asyncio-runner.rst

This commit updates the documentation for Runner.run() and
asyncio.run() in asyncio-runner.rst. In the process, I removed
documentation previously added in runners.py.

kumaraditya303 committed:
Merge branch 'main' into asyncio-runner-enhancement

kumaraditya303 committed:
add versionchanged

kumaraditya303 committed:
code -> coro
